### PR TITLE
No validation for pre-processor symbols in C# Parse Options, and report multiple errors

### LINKED ISSUE #15900: Change how pre-processor symbols are verified
Today the `CSharpParseOptions` type is inconsistent with how it verifies pre-processor symbols:

- The constructor will verify all symbols are valid identifiers and throw an exception when they are not. 
- The `WithPreProcessorSymbols` method does no verification. 

This inconsistency is bad for bot the compiler and consumers of the compiler API.  After some discussion the compiler team decided to change the API in the following way:

- The `CSharpParseOptions` type will never throw on pre-processor symbols that are invalid identifiers.
- Diagnostics will be produced for invalid identifiers in the pre-processor symbols

As a part of fixing this we should examine uses like in the following PR to see if they need to be cleaned up after the change.  

https://github.com/dotnet/roslyn/pull/15798

### LINKED ISSUE #16913: csc reports errors only one pre-processor symbol per -define flag
Example:
`csc -define:valid1,2,3`

**Expected**: Errors for invalid symbols `2` and `3`.
**Actual**: Error for just `2`.

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #15900 in both C# and VB, by removing validation from ParseOptions, and adding tests to make sure that it doesn't throw exceptions.

Fixes #16913 only in C#, as VB parses input as tokens (not by splitting input), and has special back-compat reasons for keeping this behavior.

Note to self: inspect breaking API change.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -3263,7 +3263,6 @@ override Microsoft.CodeAnalysis.CSharp.Syntax.XmlTextSyntax.Accept<TResult>(Micr
 override Microsoft.CodeAnalysis.CSharp.Syntax.YieldStatementSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor visitor) -> void
 override Microsoft.CodeAnalysis.CSharp.Syntax.YieldStatementSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
 static Microsoft.CodeAnalysis.CSharp.CSharpCommandLineParser.Default.get -> Microsoft.CodeAnalysis.CSharp.CSharpCommandLineParser
-static Microsoft.CodeAnalysis.CSharp.CSharpCommandLineParser.ParseConditionalCompilationSymbols(string value, out System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostic> diagnostics) -> System.Collections.Generic.IEnumerable<string>
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt]: Was this actually shipped?  If so, this is a binary breaking change.

**Comment by OmarTawfik** [comment]
@CyrusNajmabadi @jaredpar @dotnet/roslyn-compat-council 
For the API change here, I'd prefer to undo it if it is going to break existing scenarios.
**But** looking at `CSharpCommandLineParser`, it looks like the only other public method there is the `Parse()`, and all others are:

1. Private (to be used from the main Parse() method.
2. Take a diagnostics list that they add errors to.

Is there a reason that this specific `ParseConditionalCompilationSymbols` was made public and made to have an `out` parameter? was it by error? is there some external user?
If not, and we're OK taking this change, I'd suggest to do it to keep the public surface aligned.

**Comment by CyrusNajmabadi** [comment]
First, let's find out when that API was added.  If it was between 2015 and 2017, we can remove it.  However, if it's been shipped from before 2017, then we basically can't remove it.  Talk to @jaredpar if you want more info on our exceptionally high breaking-change bar.

File: src/Compilers/CSharp/Test/Syntax/Parsing/CSharpParseOptionsTests.cs [hunk 2]
```diff
@@ -44,15 +45,26 @@ public void ConstructorValidation()
...
         {
-            // command line gives CS2029: Invalid value for '/define'; 'xxx' is not a valid identifier
-            Assert.Throws<ArgumentException>(() => new CSharpParseOptions(preprocessorSymbols: ImmutableArray.Create<string>("")));
-            Assert.Throws<ArgumentException>(() => new CSharpParseOptions(preprocessorSymbols: ImmutableArray.Create<string>(" ")));
-            Assert.Throws<ArgumentException>(() => new CSharpParseOptions(preprocessorSymbols: ImmutableArray.Create<string>("Good", "Bad.Symbol")));
-            Assert.Throws<ArgumentException>(() => new CSharpParseOptions(preprocessorSymbols: ImmutableArray.Create<string>("123", "Good")));
-            Assert.Throws<ArgumentException>(() => new CSharpParseOptions(preprocessorSymbols: ImmutableArray.Create<string>("Good", null, @"Bad\Symbol")));
+            var options = new CSharpParseOptions(preprocessorSymbols: new string[] { "valid1", "2invalid" });
+
+            Assert.Equal(2, options.PreprocessorSymbolNames.Count());
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/CSharpParseOptionsTests.cs]: I think there is an `AssertEx.Equal` overload that could be used for this.

**Comment by OmarTawfik** [comment]
PR changes reverted, per general guidance from team.
@dotnet/roslyn-compiler  can I get a second review?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 3]
```diff
@@ -1470,7 +1470,7 @@ private ImmutableArray<string> BuildSearchPaths(string sdkDirectoryOpt, List<str
 
         public static IEnumerable<string> ParseConditionalCompilationSymbols(string value, out IEnumerable<Diagnostic> diagnostics)
         {
-            Diagnostic myDiagnostic = null;
+            List<Diagnostic> myDiagnostics = new List<Diagnostic>();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Nit: `outputDiagnostics` or `resultDiagnostics` maybe?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: >List<Diagnostic> myDiagnostics = new List<Diagnostic>(); [](start = 12, length = 56)

Can we use a DiagnosticBag instead (probably from the pool)?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Or an ArrayBuilder? We usually don't use List of T.

---
In reply to: [100169189](https://github.com/dotnet/roslyn/pull/16920#discussion_r100169189) [](ancestors = 100169189)

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb [hunk 4]
```diff
@@ -8101,6 +8101,31 @@ End Module
             Assert.Equal($"error BC2012: can't open '{sourceLinkPath}' for writing: Fake IOException{Environment.NewLine}", outWriter.ToString())
         End Sub
 
+        <WorkItem(15900, "https://github.com/dotnet/roslyn/issues/15900")>
+        <Fact>
+        Public Sub CompilingCodeWithInvalidPreProcessorSymbolsShouldErrorOut()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb]: >CompilingCodeWithInvalidPreProcessorSymbolsShouldErrorOut [](start = 19, length = 57)

Could you include how the errors look like as text, as comments perhaps?

**Comment by OmarTawfik** [comment]
> It doesn't look like we are validating pre-processor symbols in parse options. I thought the plan was to move validation elsewhere.
@jaredpar can you please check @AlekseyTs's comment here per #15900 description?

**Comment by OmarTawfik** [comment]
Based on discussion with AlekseyTs, there are surfaces that are not covered by this change.
For example `SyntaxFactory.ParseTokens()`. Closing for now.

**Pull request closed without merging**